session: include partition key in RequestSpan in human readable form

piodul opened the pull request:
This PR changes the format used by `RequestSpan` to represent the partition key for a given query. Previously, it was the same representation that the driver creates before applying MurmurHash3 to it, encoded in base64 - the format is documented so technically users can decode it themselves, but it's quite inconvenient. Now, the driver decodes the contents of the partition key and represents it in a human-readable form, similar to the CQL syntax.

For example, given the schema

```cql
create table ks.t (pk1 int, pk2 time, pk3 frozen<set<text>>, primary key(pk1, pk2, pk3));
```

and the request

```rust
session
    .execute(
        &prepared,
        (
            6622i32,
            Time(chrono::Duration::seconds(12323)),
            vec!["quick", "brown", "fox"],
        ),
    )
    .await
    .unwrap();
```

previously, the encoded key would look like this:

```
partition_key=0004000019de00000800000b352c099e0000001d0000000300000005717569636b0000000562726f776e00000003666f7800
```

but now it looks like this:

```
partition_key=6622,'03:25:23.000000000',{'quick','brown','fox'}
```

Fixes: https://github.com/scylladb/scylla-rust-driver/issues/720

## Pre-review checklist

<!--
    Make sure you took care of the issues on the list.
    Put 'x' into those boxes which apply.
    You can also create the PR now and click on all relevant checkboxes.
    See CONTRIBUTING.md for more details.
-->

- [x] I have split my patch into logically separate commits.
- [x] All commit messages clearly explain what they change and why.
- [x] I added relevant tests for new features and bug fixes.
- [x] All commits compile, pass static checks and pass test.
- [x] PR description sums up the changes and reasons why they should be introduced.
- [ ] ~I have provided docstrings for the public items that I want to introduce.~
- [ ] ~I have adjusted the documentation in `./docs/source/`.~
- [x] I added appropriate `Fixes:` annotations to PR description.

wprzytula commented on scylla/src/utils/pretty.rs:
```diff
@@ -19,3 +23,310 @@ impl<'a> UpperHex for HexBytes<'a> {
+            "{1,3,2}"
+        );
+
+        let mut tuple: Vec<_> = list_or_set.into_iter().map(Some).collect();
+        tuple.push(None);
```
Possibly cleaner:

wprzytula replied:
```suggestion
        let tuple = Vec::from_iter(list_or_set.into_iter().map(Some).chain(std::iter::once(None)));
```

wprzytula replied:
`FromIterator` should be encouraged, as it reduces verbosity induced by `collect()` (either type hint or turbofish).

piodul replied:
> `FromIterator` should be encouraged, as it reduces verbosity induced by `collect()` (either type hint or turbofish).

Why? I quite like `.collect()` to be fair. It leads to more natural flow in the code. When you work with iterators, you perform a series of steps (such as map, filter, etc.) and then finally you `.collect()` the results.

wprzytula replied:
This is a valid counterargument.

wprzytula commented on scylla/src/statement/prepared_statement.rs:
```diff
@@ -337,3 +341,132 @@ pub enum PartitionKeyError {
+            return None;
+        }
+
+        let idx = self.prepared_metadata.pk_indexes[self.index].index as usize;
```
Please rename it to `col_idx` for less ambiguity.

wprzytula commented on scylla/src/statement/prepared_statement.rs:
```diff
@@ -337,3 +341,132 @@ pub enum PartitionKeyError {
+pub(crate) struct PartitionKeyDecoder<'pk> {
+    prepared_metadata: &'pk PreparedMetadata,
+    partition_key: &'pk [u8],
+    index: usize,
```
Perhaps `value_index` or a similar name could help distinguish this conceptually from column indices, which are involved too.

wprzytula commented on scylla/src/statement/prepared_statement.rs:
```diff
@@ -337,3 +341,132 @@ pub enum PartitionKeyError {
+    }
+}
+
+impl<'pk> PartitionKeyDecoder<'pk> {
+    fn parse_cell(&mut self) -> Result<&'pk [u8], ParseError> {
+        let buf = &mut self.partition_key;
+        let len = buf.read_u16::<BigEndian>()? as usize;
+        if buf.len() < len {
+            return Err(ParseError::IoError(std::io::Error::new(
+                std::io::ErrorKind::UnexpectedEof,
+                "value too short",
+            )));
+        }
+        let col = &buf[..len];
+        *buf = &buf[len..];
+        buf.read_u8()?;
+        Ok(col)
+    }
+}
```
Please add some comments for future readers, explaining the format, or at least provide a link to the source describing it.

wprzytula commented on scylla/src/statement/prepared_statement.rs:
```diff
@@ -337,3 +341,132 @@ pub enum PartitionKeyError {
+            ks_name: "ks".to_owned(),
+            table_name: "t".to_owned(),
+        };
+        let meta = PreparedMetadata {
+            flags: 0,
+            col_count,
+            col_specs: (0..col_count)
+                .map(|i| ColumnSpec {
+                    name: format!("col_{}", i),
+                    table_spec: table_spec.clone(),
+                    typ: ColumnType::Int,
+                })
+                .collect(),
+            pk_indexes: (0..col_count)
+                .map(|i| PartitionKeyIndex {
+                    index: i as u16,
+                    sequence: i as u16,
+                })
+                .collect(),
+        };
```
Is such `PreparedMetadata` general enough to cover the most important test cases? I mean, column type is always `Int`, and sequence numbers are equal to indices. I'm not sure though if we need to test other cases _here_.

wprzytula reviewed (commented):
Looks good, only some naming/comments nits.

piodul commented:
v2:

- `std::iter::chain` instead of collect to vector then `push`,
- Rename `idx` -> `col_idx`,
- Rename `index` -> `value_index`,
- Added a link to the explanation of the format that `PartitionKeyDecoder` decodes from,
- Reworked the tests for `PartitionKeyDecoder` a bit and added a more complicated case that involves out-of-order columns and different column types.
